rustworkx-core/src/dag_algo.rs

```diff
@@ -314,6 +319,176 @@ where
+///     will be the first layer in the output
+///
+/// ```
+/// use rustworkx_core::petgraph::prelude::*;
+/// use rustworkx_core::dag_algo::layers;
+/// use rustworkx_core::dictmap::*;
+///
+/// let edge_list = vec![
+///  (0, 1),
+///  (1, 2),
+///  (2, 3),
+///  (3, 4),
+/// ];
+///
+/// let graph = DiGraph::<u32, u32>::from_edges(&edge_list);
+/// let layers: Vec<Vec<NodeIndex>> = layers(&graph, vec![0.into(),1.into()]).unwrap().collect();
+/// let expected_layers: Vec<Vec<NodeIndex>> = vec![
+///     vec![0.into(),1.into()],
+///     vec![1.into(),2.into()],
+///     vec![2.into(),3.into()],
+///     vec![3.into(),4.into()],
+///     vec![4.into()]
+/// ];
+/// assert_eq!(layers, expected_layers)
+/// ```
+pub fn layers<G>(
+    graph: G,
+    first_layer: Vec<G::NodeId>,
+) -> Result<LayersIter<G, G::NodeId>, LayersError>
+where
+    G: NodeIndexable // Used in from_index and to_index.
+        + NodeCount // Used in find_cycle
+        + EdgeCount // Used in find_cycle
+        + Visitable // Used in find_cycle
+        + IntoNodeIdentifiers // Used for .node_identifiers
+        + IntoNeighborsDirected // Used for .neighbors_directed
+        + IntoEdgesDirected, // Used for .edged_directed
+    <G as GraphBase>::NodeId: Debug + Copy + Eq + Hash,
+{
+    let cur_layer = first_layer;
+    // Throw error if a cycle is found at the current node
+    if let Some(node) = &cur_layer.first() {
+        let check_cycle = find_cycle(&graph, Some(**node));
+        if !check_cycle.is_empty() {
+            return Err(LayersError(Some(format!(
+                "Found cycle at {:?}",
+                check_cycle
+            ))));
+        }
+    }
+    // Check that the nodes exist within the graph
+    let node_ids: HashSet<G::NodeId> = graph.node_identifiers().collect();
+    for node in cur_layer.iter() {
+        if !node_ids.contains(node) {
+            panic!(
+                "An index input in 'first_layer' {:?} is not a valid node index in the graph",
+                node
+            );
+        }
+    }
+    Ok(LayersIter::new(graph, cur_layer))
+}
+
+#[derive(Debug, Clone)]
+pub struct LayersIter<G, N> {
+    graph: G,
+    cur_layer: Vec<N>,
+    next_layer: Vec<N>,
+    predecessor_count: HashMap<N, usize>,
+    first_iter: bool,
+}
+
+impl<G, N> LayersIter<G, N>
+where
+    G: NodeIndexable // Used in from_index and to_index.
+        + NodeCount // Used in find_cycle
+        + EdgeCount // Used in find_cycle
+        + Visitable // Used in find_cycle
+        + IntoNeighborsDirected // Used for .neighbors_directed
+        + IntoEdgesDirected // Used for .edged_directed
+        + GraphBase<NodeId = N>,
+    N: Debug + Copy + Eq + Hash,
+{
+    fn new(graph: G, first_layer: Vec<N>) -> Self {
+        Self {
+            graph,
+            cur_layer: first_layer,
+            next_layer: vec![],
+            predecessor_count: HashMap::new(),
+            first_iter: true,
+        }
+    }
+}
```

Comment:
Do we need a `new()` here? It feels a bit unnecessary since this is just the iterator return type from this function and not something users should be manually interacting with.

Reply:
No, it is not necessary, I was just trying to avoid having to initialize all the other empty attributes manually. This will be removed in a future commit.